File: desktop/src/shared/api/humanPolicy.ts
import { invokeTauri } from "@/shared/api/tauri";

export interface CardChoice {
  choice_id: string;
  label: string;
}

export interface CardAnswer {
  choice_id: string;
  answered_at_ms: number;
  resumed: boolean;
}

export interface HumanCard {
  card_id: string;
  turn_id: string;
  owner_id: string;
  agent_id: string;
  kind: "clarify" | "approval" | "choice";
  title: string;
  body: string;
  choices: CardChoice[];
  created_at_ms: number;
  answered: CardAnswer | null;
}

export interface CreateHumanCardInput {
  card_id: string;
  turn_id: string;
  owner_id: string;
  agent_id: string;
  kind: string;
  title: string;
  body: string;
  choices: CardChoice[];
}

export interface SpendGuardStatus {
  window_ms: number;
  max_wakes_per_window: number;
  max_runs_per_window: number;
  grace_ms: number;
  snooze_ms: number;
  wakes_in_window: number;
  runs_in_window: number;
  paused: boolean;
}

export interface SpendGuardConfig {
  window_ms: number;
  max_wakes_per_window: number;
  max_runs_per_window: number;
  grace_ms: number;
  snooze_ms: number;
}

export async function listHumanCards(): Promise<HumanCard[]> {
  return invokeTauri<HumanCard[]>("list_human_cards");
}

export async function createHumanCard(
  input: CreateHumanCardInput,
): Promise<HumanCard> {
  return invokeTauri<HumanCard>("create_human_card", { input });
}

export async function answerHumanCard(
  card_id: string,
  choice_id: string,
): Promise<HumanCard> {
  return invokeTauri<HumanCard>("answer_human_card", {
    input: { card_id, choice_id },
  });
}

export async function getSpendGuardStatus(): Promise<SpendGuardStatus> {
  return invokeTauri<SpendGuardStatus>("get_spend_guard_status");
}

export async function updateSpendGuardConfig(
  config: SpendGuardConfig,
): Promise<SpendGuardStatus> {
  return invokeTauri<SpendGuardStatus>("update_spend_guard_config", { config });
}

export async function toggleSpendGuardPause(
  paused: boolean,
): Promise<SpendGuardStatus> {
  return invokeTauri<SpendGuardStatus>("toggle_spend_guard_pause", { paused });
}

export interface SkillVersion {
  skill_id: string;
  owner_id: string;
  version: number;
  manifest_id: string;
  created_at_ms: number;
  private: boolean;
}

export interface ManifestFile {
  path: string;
  sha256: string;
}

export interface CaptureManifest {
  manifest_id: string;
  owner_id: string;
  title: string;
  summary: string;
  files: ManifestFile[];
}

export interface PreflightCheck {
  name: string;
  passed: boolean;
  detail: string;
}

export interface PreflightFrame {
  frame_id: string;
  checks: PreflightCheck[];
}

export interface DryRunRequest {
  skill_id: string;
  version: number;
  capabilities: string[];
}

export interface DryRunResult {
  skill_id: string;
  version: number;
  allowed: boolean;
  detail: string;
}

export async function listCuratedSkills(): Promise<SkillVersion[]> {
  return invokeTauri<SkillVersion[]>("list_curated_skills");
}

export async function preflightSkillCapture(params: {
  owner_id: string;
  manifest: CaptureManifest;
  frame_a: PreflightFrame;
  frame_b: PreflightFrame;
}): Promise<SkillVersion> {
  return invokeTauri<SkillVersion>("preflight_skill_capture", params);
}

export async function dryRunSkillCapability(
  req: DryRunRequest,
): Promise<DryRunResult> {
  return invokeTauri<DryRunResult>("dry_run_skill_capability", { req });
}

export type WorkspaceLifecycleState =
  | "prepared"
  | "admitted"
  | "creating"
  | "active"
  | "terminating"
  | "terminal";

export interface ViewerPresence {
  viewer_id: string;
  last_seen_unix_ms: number;
  active_view: string;
}

export interface WorkspaceFrameUpdate {
  frame_seq: number;
  payload_bytes: number;
  timestamp_unix_ms: number;
}

export interface WorkspaceObserverContract {
  workspace_id: string;
  lifecycle: WorkspaceLifecycleState;
  viewers: ViewerPresence[];
  frame_updates: WorkspaceFrameUpdate[];
  recording_enabled: boolean;
  scheduled_input_json?: string | null;
}

export async function getWorkspaceObserver(
  workspaceId: string,
): Promise<WorkspaceObserverContract> {
  return invokeTauri<WorkspaceObserverContract>("get_workspace_observer", {
    workspaceId,
  });
}

export async function setWorkspaceObserverPresence(params: {
  workspaceId: string;
  viewerId: string;
  activeView: string;
}): Promise<WorkspaceObserverContract> {
  return invokeTauri<WorkspaceObserverContract>(
    "set_workspace_observer_presence",
    params,
  );
}

export async function toggleWorkspaceRecording(params: {
  workspaceId: string;
  enabled: boolean;
}): Promise<WorkspaceObserverContract> {
  return invokeTauri<WorkspaceObserverContract>(
    "toggle_workspace_recording",
    params,
  );
}

export interface AutomationDefinition {
  definition_id: string;
  owner_id: string;
  name: string;
  revision: number;
  enabled: boolean;
  created_at_ms: number;
  updated_at_ms: number;
  config_json: unknown;
}

export type AutomationRunState = "pending" | "delivered" | "acked" | "failed";

export interface AutomationRun {
  run_id: string;
  wake_id: string;
  definition_id: string;
  owner_id: string;
  revision: number;
  state: AutomationRunState;
  attempts: number;
  created_at_ms: number;
  updated_at_ms: number;
}

export interface AutomationWake {
  wake_id: string;
  definition_id: string;
  owner_id: string;
  revision: number;
  payload_json: unknown;
  created_at_ms: number;
}

export async function listAutomationDefinitions(): Promise<
  AutomationDefinition[]
> {
  return invokeTauri<AutomationDefinition[]>("list_automation_definitions");
}

export async function createAutomationDefinition(params: {
  definitionId: string;
  name: string;
  ownerId: string;
  configJson: unknown;
}): Promise<AutomationDefinition> {
  return invokeTauri<AutomationDefinition>(
    "create_automation_definition",
    params,
  );
}

export async function toggleAutomationEnabled(params: {
  definitionId: string;
  enabled: boolean;
}): Promise<AutomationDefinition> {
  return invokeTauri<AutomationDefinition>("toggle_automation_enabled", params);
}

export async function listAutomationRuns(): Promise<AutomationRun[]> {
  return invokeTauri<AutomationRun[]>("list_automation_runs");
}

export async function triggerAutomationWake(params: {
  wakeId: string;
  definitionId: string;
  ownerId: string;
  revision: number;
  payloadJson: unknown;
}): Promise<AutomationWake> {
  return invokeTauri<AutomationWake>("trigger_automation_wake", params);
}

File: desktop/src-tauri/src/commands/human_policy.rs
use serde::{Deserialize, Serialize};
use tauri::State;
use crate::app_state::AppState;
use buzz_lifecycle::{
    CardAnswer, CardChoice, CardKind, CaptureManifest, DryRunRequest, DryRunResult,
    HumanCard, ManifestFile, PreflightCheck, PreflightFrame, SkillCurator, SkillVersion,
    SpendGuardConfig,
};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HumanCardWire {
    pub card_id: String,
    pub turn_id: String,
    pub owner_id: String,
    pub agent_id: String,
    pub kind: String,
    pub title: String,
    pub body: String,
    pub choices: Vec<CardChoiceWire>,
    pub created_at_ms: i64,
    pub answered: Option<CardAnswerWire>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CardChoiceWire {
    pub choice_id: String,
    pub label: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CardAnswerWire {
    pub choice_id: String,
    pub answered_at_ms: i64,
    pub resumed: bool,
}

impl From<HumanCard> for HumanCardWire {
    fn from(c: HumanCard) -> Self {
        Self {
            card_id: c.card_id,
            turn_id: c.turn_id,
            owner_id: c.owner_id,
            agent_id: c.agent_id,
            kind: "action_request".to_string(),
            title: c.title,
            body: c.body,
            choices: c
                .choices
                .into_iter()
                .map(|ch| CardChoiceWire {
                    choice_id: ch.choice_id,
                    label: ch.label,
                })
                .collect(),
            created_at_ms: c.created_at_ms,
            answered: c.answered.map(|a| CardAnswerWire {
                choice_id: a.choice_id,
                answered_at_ms: a.answered_at_ms,
                resumed: a.resumed,
            }),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CreateHumanCardInput {
    pub card_id: String,
    pub turn_id: String,
    pub owner_id: String,
    pub agent_id: String,
    pub kind: String,
    pub title: String,
    pub body: String,
    pub choices: Vec<CardChoiceWire>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SpendGuardStatusWire {
    pub window_ms: i64,
    pub max_wakes_per_window: u32,
    pub max_runs_per_window: u32,
    pub grace_ms: i64,
    pub snooze_ms: i64,
    pub wakes_in_window: u32,
    pub runs_in_window: u32,
    pub paused: bool,
    pub snoozed: bool,
    pub in_grace: bool,
}

#[tauri::command]
pub async fn list_human_cards(
    state: State<'_, AppState>,
) -> Result<Vec<HumanCardWire>, String> {
    let broker = state.human_card_broker.lock().map_err(|e| e.to_string())?;
    Ok(broker.all_cards().into_iter().map(HumanCardWire::from).collect())
}

#[tauri::command]
pub async fn create_human_card(
    input: CreateHumanCardInput,
    state: State<'_, AppState>,
) -> Result<HumanCardWire, String> {
    let mut broker = state.human_card_broker.lock().map_err(|e| e.to_string())?;
    let now = chrono::Utc::now().timestamp_millis();
    let card = HumanCard {
        card_id: input.card_id,
        turn_id: input.turn_id,
        owner_id: input.owner_id,
        agent_id: input.agent_id,
        kind: CardKind::ActionRequest,
        title: input.title,
        body: input.body,
        choices: input
            .choices
            .into_iter()
            .map(|c| CardChoice {
                choice_id: c.choice_id,
                label: c.label,
            })
            .collect(),
        created_at_ms: now,
        answered: None,
    };
    let created = broker.create(card).map_err(|e| format!("{:?}", e))?;
    Ok(HumanCardWire::from(created))
}

#[tauri::command]
pub async fn answer_human_card(
    card_id: String,
    choice_id: String,
    state: State<'_, AppState>,
) -> Result<HumanCardWire, String> {
    let mut broker = state.human_card_broker.lock().map_err(|e| e.to_string())?;
    let now = chrono::Utc::now().timestamp_millis();
    let answered = broker
        .answer(&card_id, &choice_id, now)
        .map_err(|e| format!("{:?}", e))?;
    Ok(HumanCardWire::from(answered))
}

#[tauri::command]
pub async fn get_spend_guard_status(
    state: State<'_, AppState>,
) -> Result<SpendGuardStatusWire, String> {
    let guard = state.spend_guard_state.lock().map_err(|e| e.to_string())?;
    let now = chrono::Utc::now().timestamp_millis();
    Ok(SpendGuardStatusWire {
        window_ms: guard.config.window_ms,
        max_wakes_per_window: guard.config.max_wakes_per_window,
        max_runs_per_window: guard.config.max_runs_per_window,
        grace_ms: guard.config.grace_ms,
        snooze_ms: guard.config.snooze_ms,
        wakes_in_window: guard.wakes_in_window,
        runs_in_window: guard.runs_in_window,
        paused: guard.is_paused(),
        snoozed: guard.is_snoozed(now),
        in_grace: guard.is_in_grace(now),
    })
}

#[tauri::command]
pub async fn update_spend_guard_config(
    config: SpendGuardConfig,
    state: State<'_, AppState>,
) -> Result<SpendGuardStatusWire, String> {
    let mut guard = state.spend_guard_state.lock().map_err(|e| e.to_string())?;
    let now = chrono::Utc::now().timestamp_millis();
    config.validate().map_err(|e| format!("{:?}", e))?;
    guard.config = config;
    Ok(SpendGuardStatusWire {
        window_ms: guard.config.window_ms,
        max_wakes_per_window: guard.config.max_wakes_per_window,
        max_runs_per_window: guard.config.max_runs_per_window,
        grace_ms: guard.config.grace_ms,
        snooze_ms: guard.config.snooze_ms,
        wakes_in_window: guard.wakes_in_window,
        runs_in_window: guard.runs_in_window,
        paused: guard.is_paused(),
        snoozed: guard.is_snoozed(now),
        in_grace: guard.is_in_grace(now),
    })
}

#[tauri::command]
pub async fn toggle_spend_guard_pause(
    paused: bool,
    state: State<'_, AppState>,
) -> Result<SpendGuardStatusWire, String> {
    let mut guard = state.spend_guard_state.lock().map_err(|e| e.to_string())?;
    let now = chrono::Utc::now().timestamp_millis();
    guard.toggle_global_pause(paused);
    Ok(SpendGuardStatusWire {
        window_ms: guard.config.window_ms,
        max_wakes_per_window: guard.config.max_wakes_per_window,
        max_runs_per_window: guard.config.max_runs_per_window,
        grace_ms: guard.config.grace_ms,
        snooze_ms: guard.config.snooze_ms,
        wakes_in_window: guard.wakes_in_window,
        runs_in_window: guard.runs_in_window,
        paused: guard.is_paused(),
        snoozed: guard.is_snoozed(now),
        in_grace: guard.is_in_grace(now),
    })
}

#[tauri::command]
pub async fn list_curated_skills(
    state: State<'_, AppState>,
) -> Result<Vec<SkillVersion>, String> {
    let curator = state.skill_curator.lock().map_err(|e| e.to_string())?;
    Ok(curator.all_skills())
}

#[tauri::command]
pub async fn preflight_skill_capture(
    owner_id: String,
    manifest: CaptureManifest,
    frame_a: PreflightFrame,
    frame_b: PreflightFrame,
    state: State<'_, AppState>,
) -> Result<SkillVersion, String> {
    let mut curator = state.skill_curator.lock().map_err(|e| e.to_string())?;
    let now = chrono::Utc::now().timestamp_millis();
    let manifest_id = manifest.manifest_id.clone();
    curator.capture(manifest).map_err(|e| format!("{:?}", e))?;
    curator
        .create_private_skill(&owner_id, &manifest_id, (&frame_a, &frame_b), now)
        .map_err(|e| format!("{:?}", e))
}

#[tauri::command]
pub async fn dry_run_skill_capability(
    req: DryRunRequest,
    state: State<'_, AppState>,
) -> Result<DryRunResult, String> {
    let curator = state.skill_curator.lock().map_err(|e| e.to_string())?;
    curator.dry_run(&req).map_err(|e| format!("{:?}", e))
}

#[tauri::command]
pub async fn get_workspace_observer(
    workspace_id: String,
    state: State<'_, AppState>,
) -> Result<buzz_workspace_controller::WorkspaceObserverContract, String> {
    let mut registry = state.workspace_observers.lock().map_err(|e| e.to_string())?;
    Ok(registry.get_or_create(&workspace_id).clone())
}

#[tauri::command]
pub async fn set_workspace_observer_presence(
    workspace_id: String,
    viewer_id: String,
    active_view: String,
    state: State<'_, AppState>,
) -> Result<buzz_workspace_controller::WorkspaceObserverContract, String> {
    let mut registry = state.workspace_observers.lock().map_err(|e| e.to_string())?;
    let now = chrono::Utc::now().timestamp_millis();
    let obs = registry.get_or_create(&workspace_id);
    obs.update_presence(viewer_id, active_view, now);
    Ok(obs.clone())
}

#[tauri::command]
pub async fn toggle_workspace_recording(
    workspace_id: String,
    enabled: bool,
    state: State<'_, AppState>,
) -> Result<buzz_workspace_controller::WorkspaceObserverContract, String> {
    let mut registry = state.workspace_observers.lock().map_err(|e| e.to_string())?;
    let obs = registry.get_or_create(&workspace_id);
    obs.set_recording(enabled);
    Ok(obs.clone())
}
